### ISSUE TITLE
shallow restore feature [design]

### ISSUE THREAD

**Issue by rjkroege** [kopia/kopia#710]
# TL;DR
I propose a feature to permit restores to be *shallow*: some entries in the restore are represented by placeholders instead of the full contents of the directory or file. These can be expanded to the full version on an as-needed basis. Snapshots of a shallow-restored tree would restore the complete contents (and be accessible in their entirety via `kopia mount`.)

# Why
Example use case: one's Linux computer has become a preemptible cloud node where the home directory is backed up via Kopia to cloud storage. (Perhaps because the original computer is in a lockdowned-office.) One wishes to economize and create the smallest cloud node for a given task and so creates a node whose disk is smaller than whole home directory. Use shallowrestore to fetch part of the home directory. Modify it. `cron`-invoked snaphots make sure that the *entire* home directory keeps getting preserved. Recreating the cloud node (say to get different node capabilities or because of preemption) doesn't require restoring the entire home directory: a series of shallow restores can complete much more quickly.

Can this be done without modifying Kopia? I don't think so. I explored a simple workflow of:

* restore repository in its totality
* put stuff in it
* snapshot tree

This was too slow to do while spinning up a new GCP node for my (not unreasonably-sized) home directory. And I had to allocate a disk larger than the smallest/cheapest tier.

I also tried fuse-mounting the repository and making a `unionfs` writeable fs on top. This eliminates the time to restore the repository (good) but the snapshot operation must still fetch the entire repository (slower than restoring) and Kopia cannot successfully (presumably locks or race conditions?) snapshot its own fuse mounts. After exploring the Kopia source, I thought that it would be easier (and run faster) to add a `shallowrestore` feature than fix self-snapshot.

# Proposed Command Details

## kopia shallowrestore (*new*)

```
kopia shallowrestore <source> <target-path>
```

Just like `kopia restore`, the source to be restored is specified in the form of a directory or file ID and optionally a sub-directory path.

For example:

```
kopia shallowrestore kffbb7c28ea6c34d6cbe555d1cf80faa9 d1’
kopia shallowrestore kffbb7c28ea6c34d6cbe555d1cf80faa9/subdir/subdir1 d1’
```

The first form will shallowrestore the contents of a `kffbb7c28ea6c34d6cbe555d1cf80faa9` to `d1`.
`d1` will be created by the restore command if it does not exist. The second form will
shallowrestore the contents of `subdir/subdir1` in `kffbb7c28ea6c34d6cbe555d1cf80faa9` to `d1`.

The important difference between this and `kopia restore` is:

* only *one layer* of a directory hierarchy is created if `<source>` is a directory
	* so a directory is made for `<source>`
	* and a placeholder is created for each entry in `<source>`
	* where a placeholder is an empty file with an extended attribute containing JSON sufficient to create the `DirEntry` corresponding to what `snapshot` would have created for a fully restored version of the placeholder.
	* *optional extension* fully populate files in the directory whose size is less than the xattr JSON or some configured size.
* if `<source>` is a file, `kopia shallowrestore` makes a file placeholder at `<target-path>`

Because of being shallow, a `kopia restore` can be *very fast*.

## Example session
Anonymized/simplified from `history` of prototype use:

```shell
$ kopia --config-file=~/proto.config   snapshot ~/projects
$ kopia --config-file=~/proto.config   snapshot list

# Mount it and browse. It should work like any other filesystem. And it did.
$ mkdir -p /tmp/kopiashallowtest
$ kopia --config-file=~/proto.config   mount kc6765f892d8ab2d44cd7e6b9910619e3 /tmp/kopiashallowtest

# /tmp/kopiashallowtest has all of ~/projects
# umount

# shallow restore
$ rm -rf ~/projects
$ kopia --config-file=~/proto.config   shallowrestore kc6765f892d8ab2d44cd7e6b9910619e3  ~/projects

# modify ~/projects with new content, move shallow placeholders to new directories
# e.g.
$ date > ~/projects/today

$ kopia --config-file=~/proto.config   snapshot  ~/projects
$ kopia --config-file=~/proto.config   mount k7e658d6745cda2084b0da9c9f4b55eed /tmp/kopiashallowtest
$ cat /tmp/kopiashallowtest/today
Sun  6 Dec 2020 06:46:31 EST

# And all remaining files that round-tripped through the `shallowrestore`/`snapshot` lifecycle were present in `/tmp/kopiashallowtest` as expected.
```

# Status
I wrote a prototype to prove to myself that this would work. Changes were reasonably small besides tests.

* `snapshot/restore/*`: refactoring to handle shallow placeholder writing
* `snapshot/snapshotfs/*`: retrieve `DirEntry` objects from placeholders while uploading a snapshot and using them in place of a hashed/walked tree.
* `tests/end_to_end_test/*`: convince myself that it works as I describe.

Are the maintainer(s) open to a PR implementing this feature? If so, I'll clean up a bit and upload a starting-point PR.

**Comment by rjkroege** [comment]
Aside: I note in passing that based on experience writing this prototype, I think that I have a good idea of how to proceed with saving/loading extended attributes (#544) and would be willing to try to add that if people are interested.

**Comment by jkowalski** [comment]
That's interesting. I was thinking of having an on-disk representation for kopia objects some time ago. 

Not sure if xattr are ideal here, because such files are not immediately discoverable - perhaps regular JSON-formatted file with a well-known extension would suffice here. (say `myfile.kopiafile` and perhaps `mydir.kopiadir`)

Note that you probably need to at least validate that object mentioned in the placeholder file actually exists in the repository, otherwise snapshot would create dangling objects making maitenance unhappy.

Definitely send us the PR, this could be really interesting. Also, I'm super interested in adding xattrs and other less-common filesystem features.

BTW. I think in the (very?) long-term we could make `mount` much smarter and allow mutations on top of existing snapshots (this would require using some form of union filesystem where the mutations are first stored on disk first, then periodically snapshotted and merged with original). There are lots of things to figure out here, but I think this would be super cool.

**Comment by rjkroege** [comment]
Thanks for the great feedback. I have some obvious next steps:

* switch from using xattr to files. Files seem more pragmatic. This shouldn't be hard.
* Upload PR for comments (I need to remove the many gratuitous calls to `Log`)

I don't understand how to best:

> at least validate that [the] object mentioned in the placeholder file actually exists

Or how I should test that I've done this properly. 1-2 sentences of hints here would probably help me do something that aligns with your thinking.

**Comment by rjkroege** [comment]
Usage notes, issues, improvements

* [ ] snapshotting of shallow trees experiences an error when there are shallow placeholders (e.g. `.kopiaignore.kopiaentry`) for ignore files. I'll think about how to address this.
* [ ] `kopia restore blah.kopiaentry blah2.kopientry` ignores `blah2.kopiaentry` (and any additional parameters.) It would be nice if expansion worked for any number of placeholders so that `kopia restore *.kopiaentry` has the desired effect. I'll fix this unless it makes the command line parsing too ugly.

**Comment by rjkroege** [comment]
Regarding shallow `.kopiaignore.kopiaentry` files: I think that the right approach is simply to *not* permit them to exist. Instead, always fully expand such a file. Theoretically, they should be reasonably small and not particularly numerous so this ought not to be a problem.

**Comment by rjkroege** [comment]
In my usage of the shallow restore feature, I'm desiring a convenience option for shallow restores to reify the directory tree but not files. This seems easy enough to add.

**Comment by julio-lopez** [comment]
@rjkroege implemented it ~2 years ago.
